File: Easy/9. Palindrome Number/solution1.js
/**
 * @param {number} x
 * @return {boolean}
 */
let isPalindrome = function (x) {
    if (x < 0 || (x % 10 == 0 && x >= 10)) {
        return false;
    }

    /**
     * 取得數字 number 的 10^index 位是什麼數字 (index >= 0)  
     * E.g. 數字 13579 的第 2 位數是 5
     * @param {number} number
     * @param {number} index
     * @returns {number}
     */
    function getDigit(number, index) {
        let result = number % 10 ** (index + 1);
        result = Math.floor(result / 10 ** index);

        return result;
    }

    // 取得數字位數
    let digits = Math.ceil(Math.log(x) / Math.log(10));

    if (digits % 2 == 1) {
        // 12321 模式
        let centerPower = (digits - 1) / 2;

        for (let shift = 1; centerPower + shift < digits; shift++) {
            if (getDigit(x, centerPower - shift) != getDigit(x, centerPower + shift)) {
                return false;
            }
        }
    } else {
        // 123321 模式
        let leftCenterPower = digits / 2;

        for (let shift = 0; leftCenterPower + shift < digits; shift++) {
            if (getDigit(x, leftCenterPower - shift - 1) != getDigit(x, leftCenterPower + shift)) {
                return false;
            }
        }
    }

    return true;
};



// ---------------------------
console.log(isPalindrome(121));
console.log(isPalindrome(-121));
console.log(isPalindrome(10));

File: Easy/9. Palindrome Number/solution1.rs
impl Solution {
    pub fn is_palindrome(x: i32) -> bool {
        fn get_digit(number: i32, index: i32) -> i32 {
            return number.to_string().chars().nth(index as usize).unwrap() as i32 - '0' as i32;
        }

        if x < 0 || (x % 10 == 0 && x >= 10) {
            return false;
        }

        let digits = (x as f64).log10().ceil() as i32;

        if digits % 2 == 1 {
            // 1234321
            let center_power: i32 = ((digits - 1) as f32 / 2.0) as i32;
            let mut shift: i32 = 1;

            while center_power + shift < digits {
                let left: i32 = get_digit(x, center_power + shift);
                let right: i32 = get_digit(x, center_power - shift);

                if left != right {
                    return false;
                }

                shift += 1;
            }
        } else {
            // 12344321
            let center_power: i32 = (digits as f32 / 2.0) as i32;
            let mut shift: i32 = 0;

            while center_power + shift < digits {
                let left: i32 = get_digit(x, center_power + shift);
                let right: i32 = get_digit(x, center_power - shift - 1);

                if left != right {
                    return false;
                }

                shift += 1;
            }
        }

        return true;
    }
}

// --------------------
struct Solution;

fn main() {
    println!("{}", Solution::is_palindrome(121));
    println!("{}", Solution::is_palindrome(-121));
    println!("{}", Solution::is_palindrome(10));
    println!("{}", Solution::is_palindrome(313));
    println!("{}", Solution::is_palindrome(1410110141));
    println!("{}", Solution::is_palindrome(1799999971));
}
